### PR TITLE
Using Sonar Metadata in FOV/Pointcloud Scripts -Ver2

### PR DESCRIPTION
**Pull request by micat001**
Marc botched the first integration when trying to update to incorporate changes to main. Instead of fixing that mess, let's just start over. All comments from https://github.com/apl-ocean-engineering/sonar_image_proc/pull/33 still apply.

## Change Overview
Biggest change is adding the sonar_msg_metadata class to unify how we access metadata across the FOV and point cloud scripts.

Other changes:
- Vectorizing geometry creation
- Clarifying comments around sonar_thresholding
- Separate self.intensity_threshold from it's intended purpose and prevent it from being used in multiple places
- Moving geometry creation (point cloud) and vector list creation (fov) back into the class method, as I think they are specific to those classes and not specific to msg metadata
- Consistent naming between the two files (mostly ```sonar_msg_*``` for the callback and msg names)

## Scaling Stuff
Here's a visualization of the first 1000 points of the sonar_msg and the different scaling methods.

The "clipped" line is what we are currently using:
image

### With no thresholding
image

### With thresholding (only positive values)
image

### With thresholding (only values > 0.1)
image

### CODE REVIEW

File: scripts/sonar_pointcloud.py [hunk 1]
```diff
@@ -143,29 +151,31 @@ def process_intensity_array(self, image: SonarImageData):
...
+        v = np.log(np.maximum(1, new_intensites)) / np.log(
+            np.iinfo(data_type).max)
 
-        v = np.clip(v, a_min=0.0, a_max=1.0)
-        intensities = (np.iinfo(np.uint8).max * v).astype(np.uint8)
+        v_max = 1.0
+        v_scaled = (v - self.cmin) / (v_max - self.cmin)
+        v_clipped = np.clip(v_scaled, a_min=0.0, a_max=1.0)
 
+        intensities = (np.iinfo(np.uint8).max * v_clipped).astype(np.uint8)
```
Comment (micat001) [inline scripts/sonar_pointcloud.py:166]: @lauralindzey  - I have tried this every which way and I keep writing it out and it ends up being what I had done earlier in the week, with two params for scale and threshold.

If the thresholding doesn't make sense here, I am happy to integrate a commit from you that has a threshold that feels more intuitive or consistent to you.

File: python/sonar_image_proc/sonar_msg_metadata.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        xx = np.array([dir.x for dir in sonar_image_msg.beam_directions])
+        yy = np.array([dir.y for dir in sonar_image_msg.beam_directions])
+        zz = np.array([dir.z for dir in sonar_image_msg.beam_directions])
+        self.azimuths = np.arctan2(-1 * yy, np.sqrt(xx**2 + zz**2))
+        self.min_azimuth = np.min(self.azimuths)
+        self.max_azimuth = np.max(self.azimuths)
+
+        elev_beamwidth = np.median(sonar_image_msg.ping_info.tx_beamwidths)
+
+        self.elevations = np.linspace(-0.5 * elev_beamwidth,
```
Comment (lauralindzey) [inline python/sonar_image_proc/sonar_msg_metadata.py:40]: This member variable is no longer used.

File: scripts/sonar_fov.py [hunk 3]
```diff
@@ -1,67 +1,38 @@
...
-import os
 import rospy
-import rospkg
-import typing
 
-from std_msgs.msg import Header
 from acoustic_msgs.msg import ProjectedSonarImage
+from std_msgs.msg import Header
 from visualization_msgs.msg import Marker, MarkerArray
 from geometry_msgs.msg import Point
```
Comment (lauralindzey) [inline scripts/sonar_fov.py:13]: alphabetization

File: scripts/sonar_pointcloud.py [hunk 4]
```diff
@@ -91,50 +101,48 @@ def __init__(self):
...
+        # cmin is a color scaling parameter. The log-scaled and normed intensity values
+        # get scaled to (cmin, 1.0)
+        self.cmin = rospy.get_param("~cmin", 0.74)
+
+        # threshold is the publish parameter if publish_all_points is false
+        self.threshold = rospy.get_param("~threshold", 0)
+
+        # Number of planes to add to the pointcloud projection.
+        # Steps are evenly distributed between the max and min elevation angles.
         self.elev_steps = rospy.get_param("~elev_steps", 2)
```
Comment (lauralindzey) [inline scripts/sonar_pointcloud.py:121]: It seems like this is unused (though I liked that feature).

Comment (micat001) [reply scripts/sonar_pointcloud.py:121]: Assuming this comment was about the ```elev_steps``` parameter.

Comment (lauralindzey) [reply scripts/sonar_pointcloud.py:121]: OK, and I finished adding it back (replacing params for min/max elevation to display)

File: scripts/sonar_pointcloud.py (hunk 1 above)
Comment (lauralindzey) [reply scripts/sonar_pointcloud.py]: Yeah, this is functionally the same as earlier in the week.

I just pushed a commit that also uses two params for colormap scale and intensity threshold, but where they are independent s.t. changing cmin doesn't affect which points are shown.

File: scripts/sonar_fov.py (hunk 3 above)
Comment (lauralindzey) [reply scripts/sonar_fov.py]: fixed.

File: python/sonar_image_proc/sonar_msg_metadata.py (hunk 2 above)
Comment (lauralindzey) [reply python/sonar_image_proc/sonar_msg_metadata.py]: fixed.

**Pull request merged**